elastic/kibana
# Prevent Task Manager poller from constantly polling without delay

Resovles https://github.com/elastic/kibana/issues/172214

In this PR, I'm making the following improvements to prevent Task Manager from constantly hitting Elasticsearch with update by query requests:
- Preventing the backpressure mechanism from changing the `pollInterval` to anything greater than 60s
- Disabling the backpressure mechanism when the `pollInterval` configured in `kibana.yml` is higher than 60s (could be implemented better but trying to keep the code small)
- Preventing the task poller from changing the `pollInterval` to a `null` or `NaN` value and logging a stack trace if ever it happens so we can better understand the issue.

## To verify

1. Turn on Task Manager debug logging by setting the following in your `kibana.yml` file.
```
logging:
  loggers:
    - name: plugins.taskManager
      level: debug
```

>Preventing the backpressure mechanism from changing the `pollInterval` to anything greater than 60s

2. Set the following custom code to trigger the backoff mechanism that will increase the poll interval
```
diff --git a/x-pack/plugins/task_manager/server/task_store.ts b/x-pack/plugins/task_manager/server/task_store.ts
index acc12b39ec0..6670afbc29b 100644
--- a/x-pack/plugins/task_manager/server/task_store.ts
+++ b/x-pack/plugins/task_manager/server/task_store.ts
@@ -14,6 +14,7 @@ import { SavedObjectError } from '@kbn/core-saved-objects-common';

 import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
 import type { SavedObjectsBulkDeleteResponse, Logger } from '@kbn/core/server';
+import { SavedObjectsErrorHelpers } from '@kbn/core/server';

 import {
   SavedObject,
@@ -136,6 +137,10 @@ export class TaskStore {
       // The poller doesn't need retry logic because it will try again at the next polling cycle
       maxRetries: 0,
     });
+
+    setInterval(() => {
+      this.errors$.next(SavedObjectsErrorHelpers.createTooManyRequestsError('', ''));
+    }, 5000);
   }

   /**
```
3. Notice the logs `Poll interval configuration changing from X to Y after seeing Z "too many request" and/or "execute [inline] script" error(s)` where Y keeps increasing and with time will stop at `60000` and not surpass it. This will take a few minutes of Kibana running to reach the limit. Once you see `Task poller now using interval of 60000ms` you will no longer see these logs because the poll interval has stopped increasing. You can wait a minute to be certain.

>Disabling the backpressure mechanism when the pollInterval configured in kibana.yml is higher than 60s (could be implemented better but trying to keep the code small)

4. Set `xpack.task_manager.poll_interval: 60000` in your `kibana.yml` file

5. Notice the logs from step 3 aren't happening but you still see max workers reduce over time.

>Preventing the task poller from changing the `pollInterval` to a `null` or `NaN` value and logging a stack trace if ever it happens so we can better understand the issue.

6. Set the following code that forces a null in the observable pipeline (still unknown how the code does this but we're testing the defensive code in a different module)
```
diff --git a/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts b/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts
index 52e11e431ba..da761a46555 100644
--- a/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts
+++ b/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts
@@ -134,7 +134,7 @@ function createPollIntervalScan(logger: Logger, startingPollInterval: number) {
         );
       }
     }
-    return newPollInterval;
+    return null;
   }, startingPollInterval);
 }
```

7. Notice `Error: Expected the new interval to be a number > 0, received: null but poller will keep using: 65000` error in the logs with a stack trace.

8. Remove the custom `xpack.task_manager.poll_interval` configuration from your `kibana.yml` file

9. Set the following code that forces the outcome of a formula to be NaN
```
diff --git a/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts b/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts
index 52e11e431ba..104997505c0 100644
--- a/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts
+++ b/x-pack/plugins/task_manager/server/lib/create_managed_configuration.ts
@@ -106,6 +106,7 @@ function createPollIntervalScan(logger: Logger, startingPollInterval: number) {
         Math.ceil(previousPollInterval * POLL_INTERVAL_INCREASE_PERCENTAGE),
         Math.max(PREFERRED_MAX_POLL_INTERVAL, startingPollInterval)
       );
+      newPollInterval = NaN;
       if (Number.isNaN(newPollInterval) || newPollInterval == null || newPollInterval < 0) {
         logger.error(
           `Poll interval configuration had an issue calculating the new poll interval: Math.min(Math.ceil(${previousPollInterval} * ${POLL_INTERVAL_INCREASE_PERCENTAGE}), Math.max(${PREFERRED_MAX_POLL_INTERVAL}, ${startingPollInterval})) = ${newPollInterval}`
```

10. Observe we log the formula as an error when the outcome is NaN
```
Poll interval configuration had an issue calculating the new poll interval: Math.min(Math.ceil(null * 1.2), Math.max(60000, 3000)) = NaN
```
